pir/cpp/context.h

```diff
@@ -0,0 +1,114 @@
+//
+// Copyright 2020 the authors listed in CONTRIBUTORS.md
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#ifndef PIR_CONTEXT_H_
+#define PIR_CONTEXT_H_
+
+#include "seal/seal.h"
+#include "util/statusor.h"
+
+namespace pir {
+
+using ::private_join_and_compute::StatusOr;
+
+class PIRContext {
+ public:
+  /**
+   * Creates a new context
+   * @returns TODO
+   **/
+  static StatusOr<std::unique_ptr<PIRContext>> Create();
+
+  /**
+   * Creates a new context from existing params
+   * @param[in] params Serialized PIR parameters
+   * @returns TODO
+   **/
+  static StatusOr<std::unique_ptr<PIRContext>> CreateFromParams(
+      const std::string& params);
+
+  /**
+   * Returns the serialized PIR parameters
+   * @returns TODO
+   **/
+  std::string SerializeParams() const;
+
+  /**
+   * Encodes a vector to a Plaintext
+   * @param[in] in Array to be encoded
+   * @returns TODO
+   **/
+  StatusOr<seal::Plaintext> Encode(const std::vector<uint64_t>& in);
+  /**
+   * Decodes a plaintext to a vector
+   * @param[in] in Plaintext to be decoded
+   * @returns TODO
+   **/
+  StatusOr<std::vector<uint64_t>> Decode(const seal::Plaintext& in);
+
+  /**
+   * Encodes, encrypts and serializes a vector
+   * @param[in] in Vector to be encrypted
+   * @returns TODO
+   **/
+  StatusOr<std::string> Encrypt(const std::vector<uint64_t>& in);
+
+  /**
+   * Deserializes, decrypts and decodes a vector
+   * @param[in] in Serialized ciphertext
+   * @returns TODO
+   **/
+  StatusOr<std::vector<uint64_t>> Decrypt(const std::string& in);
+
+  /**
+   * Serializes a ciphertext
+   * @param[in] in Ciphertext to be serialized
+   * @returns TODO
+   **/
+  StatusOr<std::string> Serialize(const seal::Ciphertext&);
+  /**
+   * Deserializes a ciphertext
+   * @param[in] in Serialized ciphertext
+   * @returns TODO
+   **/
+  StatusOr<seal::Ciphertext> Deserialize(const std::string& in);
+
+  /**
+   * Returns an Evaluator instance
+   * @returns TODO
+   **/
+  std::shared_ptr<seal::Evaluator>& Evaluator();
+
+ private:
+  PIRContext(const seal::EncryptionParameters&);
+
+  static seal::EncryptionParameters generateEncryptionParams(
+      uint32_t poly_modulus_degree = 4096, uint32_t plain_modulus = 1032193);
+
+  seal::EncryptionParameters parms_;
```

Comment:
We're going to need more than just the encryption parameters here. We'll also need the number of elements and the maximum size of each element at the very least.